Add HCI driver generic to NXP platforms

yeaissa opened the pull request:
- Add NXP HCI driver
- Add yaml binding file for NXP HCI
- Update host configuration for NXP platforms

axelnxp committed:
modules: Add Kconfig to Kconfig.mcux for BLE support

Added NXP_FW_LOADER for NXP platforms.
Added NXP_RF_IMU for NXP platforms.
Added MONOLITHIC Flags for NXP platforms.

Signed-off-by: Axel Le Bourhis <axel.lebourhis@nxp.com>

axelnxp committed:
drivers: hci: Add NXP HCI driver

Add HCI driver generic to NXP platforms.
Update west.yml to have ble support for rw61x

Signed-off-by: Axel Le Bourhis <axel.lebourhis@nxp.com>
Signed-off-by: Yassine El Aissaoui <yassine.elaissaoui@nxp.com>

yeaissa committed:
dts: binding: add yaml file for nxp HCI

- used to define hci info

Signed-off-by: Yassine EL -AISSAOUI <yassine.elaissaoui@nxp.com>

jori-nordic commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -132,6 +133,7 @@ config BT_RX_STACK_SIZE
 	default 2600 if BT_MESH
 	default 2048 if BT_AUDIO
 	default 2200 if BT_SETTINGS
+	default 1536 if BT_NXP && DEBUG
```
why the increase? the hci driver shouldn't have an influence on the host RX stack.

yeaissa replied:
Indeed, The increase is not needed. Corrected

decsny commented on dts/bindings/arm/nxp,hci-ble.yaml:
```diff
@@ -0,0 +1,12 @@
+# Copyright 2023 NXP
+# SPDX-License-Identifier: Apache-2.0
+
+description: HCI interrupt info
```
since the binding  name is generically "hci-ble" the description should match this or the name should match the description

yeaissa replied:
Corrected,Thanks.

decsny reviewed (dismissed):
this should be enabled for a platform in this PR so that it is tested

yeaissa committed:
soc: rw61x: Add BLE support for rw61x

- Add SMU regions
- Add HCI definition
- Add config when BT is enabled

Signed-off-by: Yassine El Aissaoui <yassine.elaissaoui@nxp.com>

yeaissa committed:
boards: rd_rw612_bga: Update board information

- Add pinctrl for flexcomm0 in usart mode
needed for BLE tester application
- Reserve the firmware storage partition
- Mark HCI as a wakeup source

Signed-off-by: Yassine El Aissaoui <yassine.elaissaoui@nxp.com>

yeaissa commented:
> this should be enabled for a platform in this PR so that it is tested

RW61x platform support added to the PR. @decsny @jori-nordic

decsny reviewed (changes requested):
please put the binding commit before the driver commit, and squash the west.yml change into the driver commit so that the next commits where you enable the features on the platform works as they are added

yeaissa commented:
> please put the binding commit before the driver commit, and squash the west.yml change into the driver commit so that the next commits where you enable the features on the platform works as they are added

Addressed. Thanks

jhedberg commented on drivers/bluetooth/hci/Kconfig.nxp:
```diff
@@ -0,0 +1,29 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+config HCI_NXP_SET_BDADDR
+	bool "HCI Set BT Address"
+	help
+	  If enabled, the BD_ADDR will be set by the host at HCI init, using a HCI vendor command.
```
This option doesn't seem to be used anywhere - maybe it should be removed?

yeaissa replied:
Indeed, this isn't used for now. Removed

jhedberg commented on samples/bluetooth/peripheral_ots/sample.yaml:
```diff
@@ -7,6 +7,7 @@ tests:
     platform_allow:
       - qemu_cortex_m3
       - qemu_x86
+      - rd_rw612_bga
```
I know it was mentioned earlier that you need more coverage, but I don't think it's right to add the board to absolutely every single sample application. We're not doing that for most HCI drivers, so I don't think we should do it for this new one either. You *should* have some coverage though, ideally so that you cover all the possible build configurations for the driver. These samples will not get exercised at runtime by upstream CI for this board, so the key thing is really the build time coverage - enabling new samples which don't increase the coverage is not helpful (just increases CI completion time without any benefit)

yeaissa replied:
Most of the samples weren't built by the CI since they don't belong to the "integration platforms" list. Your comment still makes sense, I have removed the platform name from all the samples/tests except peripheral_ht, central_ht and tester application.
Thanks a lot.

jhedberg commented on dts/arm/nxp/nxp_rw6xx_common.dtsi:
```diff
@@ -361,6 +395,12 @@
 		};
 	};
 
+	hci: hci_ble {
+		compatible = "nxp,hci-ble";
```
Just a reminder for the future: This node and associated binding will need to be converted to a real HCI driver instance node along with #72323

yeaissa replied:
Thanks for the information @jhedberg. When is this expected on the mainline ? I'll work on the adoption of this rework when it's ready.

jhedberg replied:
> Thanks for the information @jhedberg. When is this expected on the mainline ? I'll work on the adoption of this rework when it's ready.

Not sure. Remains to be seen if it'll make it before the next release or if it'll only go in after it. Anyway, it shouldn't prevent new HCI drivers from being merged meanwhile (it just means that the list of drivers needing converting keeps growing)

yeaissa committed:
samples: tests: bluetooth: Add support for rd_rw612_bga

Adding bt_tester support.

Adding bt_tester app build
with Twister for rd_rw612_bga board

Fix failing tests/samples due to unsufficent MPU ressources
by releasing unnecessary MPU regions.

Disable twister build on samples that are not meant to
be supported on this board, so exclude it from the test cases.

Update peripheral/central ht sample.yaml files
to test build for rd_rw612_bga.

Signed-off-by: Axel Le Bourhis <axel.lebourhis@nxp.com>
Signed-off-by: Yassine El Aissaoui <yassine.elaissaoui@nxp.com>

yeaissa commented:
@jhedberg could you please check again ? Thanks

jhedberg commented on drivers/bluetooth/hci/hci_nxp.c:
```diff
@@ -0,0 +1,421 @@
+#define HCI_WAKEUP_IRQ_N DT_INST_IRQ_BY_NAME(0, wakeup_int, irq)
+#define HCI_WAKEUP_IRQ_P DT_INST_IRQ_BY_NAME(0, wakeup_int, priority)
+
+#define HCI_CMD 0x01
+#define HCI_ACL 0x02
+#define HCI_EVT 0x04
```
Seems like these could use the `BT_HCI_H4_*` defines from `<zehyr/bluetooth/hci_types.h>`

dleach02 commented:
@jhedberg can you clarify what your current change request is on this PR? We want to try to bring this to closure so we can merge and prepare for the next dependent PR.

yeaissa commented:
@jhedberg, rebased and addressed your comment
The build failure doesn't seem to be linked to the PR changes

dleach02 commented:
@yeaissa I rebased your branch to fix a merge conflict.

yeaissa commented:
Please lets move on with this PR, if there is nothing else to address. @jhedberg

dleach02 commented:
@jori-nordic, @jhedberg, I merged the HAL which required the west update...
